k2/csrc/default_context.cu

```diff
@@ -52,65 +53,126 @@ class CpuContext : public Context {
+        pinned_context->CopyDataTo(num_bytes, region->data, dst_context, dst);
+        break;
+      }
+      default:
+        K2_LOG(FATAL) << "Unsupported device type: " << device_type;
+        break;
+    }
+  }
 };
 
 class CudaContext : public Context {
  public:
   explicit CudaContext(int32_t gpu_id) : gpu_id_(gpu_id) {
-    if (gpu_id_ != -1) {
+#ifdef K2_WITH_CUDA
+    if (gpu_id != -1) {
       auto ret = cudaSetDevice(gpu_id_);
       K2_CHECK_CUDA_ERROR(ret);
+    } else {
+      int current_gpu_id;
+      auto ret = cudaGetDevice(&current_gpu_id);
+      K2_CHECK_CUDA_ERROR(ret);
+      gpu_id_ = current_gpu_id;
     }
-    // TODO(haowen): choose one from available GPUs if gpu_id == -1?
-    // and handle GPU ids from multiple machines.
-    auto ret = cudaStreamCreate(&stream_);
-    K2_CHECK_CUDA_ERROR(ret);
+    allocator_ = new cub::CachingDeviceAllocator();
+#else
+    K2_LOG(FATAL) << "Unreachable code.";
+#endif
   }
-  ContextPtr GetCpuContext() override { return k2::GetCpuContext(); }
   DeviceType GetDeviceType() const override { return kCuda; }
   int32_t GetDeviceId() const override { return gpu_id_; }
 
   void *Allocate(std::size_t bytes, void **deleter_context) override {
     void *p = nullptr;
-    if (bytes) {
-      auto ret = cudaMalloc(&p, bytes);
-      K2_CHECK_CUDA_ERROR(ret);
-    }
+#ifdef K2_WITH_CUDA
+    DeviceGuard guard(gpu_id_);
+    // the default stream is 0
+    auto ret = allocator_->DeviceAllocate(&p, bytes);
+    K2_CHECK_CUDA_ERROR(ret);
     if (deleter_context != nullptr) *deleter_context = nullptr;
+#endif
     return p;
   }
 
+  void CopyDataTo(size_t num_bytes, const void *src, ContextPtr dst_context,
+                  void *dst) override {
+    DeviceType device_type = dst_context->GetDeviceType();
+    switch (device_type) {
+      case kCpu: {
+        cudaError_t ret =
+            cudaMemcpy(dst, src, num_bytes, cudaMemcpyDeviceToHost);
+        K2_CHECK_CUDA_ERROR(ret);
+        break;
+      }
+      case kCuda: {
+        DeviceGuard guard(dst_context);
+        cudaError_t ret =
+            cudaMemcpyAsync(dst, src, num_bytes, cudaMemcpyDeviceToDevice,
+                            dst_context->GetCudaStream());
+        K2_CHECK_CUDA_ERROR(ret);
+        break;
+      }
+      default:
+        K2_LOG(FATAL) << "Unsupported device type: " << device_type;
+        break;
+    }
+  }
+
   bool IsCompatible(const Context &other) const override {
     return other.GetDeviceType() == kCuda && other.GetDeviceId() == gpu_id_;
   }
 
   void Deallocate(void *data, void * /*deleter_context*/) override {
-    auto ret = cudaFree(data);
+#ifdef K2_WITH_CUDA
+    DeviceGuard guard(gpu_id_);
+    auto ret = allocator_->DeviceFree(data);
     K2_CHECK_CUDA_ERROR(ret);
+#endif
   }
 
   cudaStream_t GetCudaStream() const override {
-    return g_stream_override.OverrideStream(stream_);
+#ifdef K2_WITH_CUDA
+    return g_stream_override.OverrideStream(0);
```

Comment:
Here, in `Pytorch_Context.cu`, we get the stream by `c10::cuda::getCurrentCudaStream`, I read the source code of Pytorch, if we don't `setCurrentCudaStream` the return stream is always the default stream (i.e. stream 0).

So, can we use the default stream 0 here? (0 might not be good here, we can add a macro some like `kDefaultStream`)